### PR TITLE
Remove QueryParseContext

### PR DESCRIPTION
**Pull request by cbuescher**
QueryParseContext is currently only used as a wrapper for an XContentParser, so
this change removes it entirely and changes the appropriate APIs that use it so
far to only accept a parser instead.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java [hunk 1]
```diff
@@ -53,7 +53,7 @@
     public static final ParseField IGNORE_UNMAPPED = new ParseField("ignore_unmapped");
     public static final QueryBuilder DEFAULT_INNER_HIT_QUERY = new MatchAllQueryBuilder();
 
-    private static final ObjectParser<InnerHitBuilder, QueryParseContext> PARSER = new ObjectParser<>("inner_hits", InnerHitBuilder::new);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java]: ❤️

File: core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java [hunk 2]
```diff
@@ -84,9 +83,10 @@ public String getName() {
     }
 
     /**
-     * Internal: Registers sub-factories with this factory. The sub-factory will be
-     * responsible for the creation of sub-aggregators under the aggregator
-     * created by this factory. This is only for use by {@link AggregatorFactories#parseAggregators(QueryParseContext)}.
+     * Internal: Registers sub-factories with this factory. The sub-factory will
+     * be responsible for the creation of sub-aggregators under the aggregator
+     * created by this factory. This is only for use by
+     * {@link AggregatorFactories#parseAggregators(org.elasticsearch.common.xcontent.XContentParser)}.
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java]: I think it is ok to import `XContentParser` to make the javadoc shorter if you like.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java [hunk 3]
```diff
@@ -57,15 +57,15 @@
     private String separator = DEFAULT_SEPARATOR;
 
     public static Aggregator.Parser getParser() {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java]: Can this be
```
public static final Aggregator.Parser PARSER;
static {
  ObjectParser objectParser = new ObjectParser();
  // build the parser
  PARSER = (name, parser) -> {
    // Stuff
  };
}
```

I think that'd be closer to how we parse other things. Not the same, but closer.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java [hunk 4]
```diff
@@ -62,43 +62,43 @@
...
-        parser.declareObject(SignificantTermsAggregationBuilder::backgroundFilter,
+        PARSER.declareObject(SignificantTermsAggregationBuilder::backgroundFilter,
                 (p, context) -> parseInnerQueryBuilder(p),
                 SignificantTermsAggregationBuilder.BACKGROUND_FILTER);
 
-        parser.declareField((b, v) -> b.includeExclude(IncludeExclude.merge(v, b.includeExclude())),
+        PARSER.declareField((b, v) -> b.includeExclude(IncludeExclude.merge(v, b.includeExclude())),
                 IncludeExclude::parseInclude, IncludeExclude.INCLUDE_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_OR_STRING);
 
-        parser.declareField((b, v) -> b.includeExclude(IncludeExclude.merge(b.includeExclude(), v)),
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: Same question as for the adjacency matrix one.

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
Yes!

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 5]
```diff
@@ -276,6 +275,9 @@ public final QueryBuilder rewrite(QueryRewriteContext queryShardContext) throws
         return rewritten;
     }
 
+    /**
+     * @throws IOException exception on rewrite
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: nit: if we add javadocs maybe we should also say what the method is for?

**Review by javanna — APPROVED** [review]
LGTM thanks @cbuescher for taking care of this

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java]: to be honest, I don't know the reason why this is constructed the way it is. Will see if I can change this.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java (hunk 4 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: This is more tricky to untangle because of the significanceHeuristicParserRegistry that needs to be passed in. No idea why, but I tried changing this but it gets a bit to involved to do it in this PR I think.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java (hunk 3 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java]: I changed this to a more "standard" approach of constructing the parser.

File: core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java (hunk 5 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: I removed this docs, they serve no real purpose

**Comment by cbuescher** [comment]
@leasticmachine test this please

**Comment by cbuescher** [comment]
@elasticmachine test this again please

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java (hunk 4 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: Fair enough. I wouldn't change the capitalization though.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: I had a quick look and opened #25519 with what I imagine the strategy is. It certainly looks big enough to be worth doing in its own PR.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java [hunk 6]
```diff
@@ -56,26 +55,23 @@
...
-            public AggregationBuilder parse(String aggregationName, QueryParseContext context) throws IOException {
-                AdjacencyMatrixAggregationBuilder result = parser.parse(context.parser(),
-                        new AdjacencyMatrixAggregationBuilder(aggregationName), context);
-                result.checkConsistency();
-                return result;
-            }
-        };
+    private static final ObjectParser<AdjacencyMatrixAggregationBuilder, Void> PARSER = new ObjectParser<>(
+            AdjacencyMatrixAggregationBuilder.NAME);
+    static {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java]: 👍  Much more like the others. nice. And less lines!

**Review by nik9000 — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java (hunk 4 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: > I wouldn't change the capitalization though

Thats done because of naming conflicts I otherwise get in https://github.com/elastic/elasticsearch/pull/25486/files#diff-e5532441d96af10726cd20f6b653e3e9R101
I could rename the XContentparser there I guess, but we user capital case for ObjectParser quiet often.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: I wouldn't name it in capital case because it isn't a constant. Otherwise I'm fine with whatever rename you like.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/significant/SignificantTermsAggregationBuilder.java]: That makes sense, sorry it took me so long to understand what you mean, but I'm so used to see these parsers as constants (because they usually are) that I didn't see this one isn't.
